### ISSUE TITLE
Document intended semantics behavior on each Flutter widget - whether standalone or building block

### ISSUE THREAD

**Issue by marcianx** [flutter/flutter#128708]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

**Context:** I ran into `TextField` merging its semantics into the parent, resulting into an incorrect semantics tree. This has already been filed as an issuea at: go/flutter-issues/26336.

**Issue:** One of the challenges I run into with Flutter component documentation is that it's not clear for each widget whether it is intended to be
- **standalone** (and provides `Semantics(container: true)`) or
- a **building block** (provides `Semantics`, but without `container: true`, intending to be merged with its parent)

This has to be determined by trial and error or digging to various abstractions in the code. Sometimes it's clear from the examples given, but this is confusing for cases like `IconButton` whose behavior depends on whether `theme.useMaterial3` is enabled:
* `theme.useMaterial3=true`: [emits `container: true` via another widget](https://github.com/flutter/flutter/blob/353b8bc87d297295069ae1b47885920f07d16e9e/packages/flutter/lib/src/material/icon_button.dart#L712-L719)
* `theme.useMaterial3=true`: [doesn't emit `container: true`](https://github.com/flutter/flutter/blob/353b8bc87d297295069ae1b47885920f07d16e9e/packages/flutter/lib/src/material/icon_button.dart#L772-L775)

One worrisome thing is that when these are set unintentionally one way or another, I'm not sure how easily they can be fixed after the fact without potential accessibility breakage (which is not visually obvious): e.g., if users use `TextField` or `IconButton` as building-blocks, but if the flutter team didn't intend them as such, then fixing the bug so that they become stand-alone might break existing semantics.

### Expected results

N/A

### Actual results

N/A

### Code sample

N/A

### Screenshots or Video

N/A

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

**Comment by darshankawar** [comment]
Thanks for the report @marcianx 
Can you provide us an example (code implementation) that shows the highlighted issue, that way we can properly address this issue.

**Comment by marcianx** [comment]
@darshankawar Could you elaborate what you need from me? I'm raising concerns about the documentation, pointed out [a related issue](https://github.com/flutter/flutter/issues/26336) that the confusion manifests as and I've pointed to various parts of the code for `IconButton` as well and the consequences of this.

**Comment by darshankawar** [comment]
Thanks for the feedback @marcianx. I was wondering if any code sample could be provided that shows the current behavior you are seeing, so that it would be useful for the team.

**Comment by marcianx** [comment]
1. The code sample in https://github.com/flutter/flutter/issues/128709 points to this issue in a Dart comment. Simplifying:
    ```dart
    Column(
      children: [
        RichText(text: TextSpan(text: 'Foo')),
        const TextField(maxLines: null, minLines: 8),
      ],
    )
    ```
results in a single `SemanticsNode` with `textField: true` _and_ semantics label `Foo` instead of having separate nodes for `RichText` and `TextField`. It seems that both of these widgets lack `container: true`. As noted, the pre-existing issue for `TextField` is https://github.com/flutter/flutter/issues/26336. The intended behavior (whether these are intended to be stand-alone or building-block widgets) should be documented.

2. For `IconButton`'s inconsistency, I saw the instances where my team had done:
    ```dart
    CustomTooltip(
      child: IconButton(icon: Icon(Icons.Add)), onPressed: () {}),
    )
    ```
    You can replace `CustomTooltip` with `Semantics(tooltip: 'Foo', child: ...)`.
    This was incorrect because with `theme.useMaterial3=true`, `IconButton` emits `Semantics(container: true, ...)` and the tooltip generates its own semantics. So I wrote a test to catch it:
    ```dart
    expect(
      tester.semantics.simulatedAccessibilityTraversal().map((node) =>
          (node.label, button: node.hasFlag(SemanticsFlag.isButton))),
      [('Add', button: true)],
    );
    ```
    but the test didn't fail like it should have. That's because the test didn't set `theme.useMaterial3=true`, and thus `IconButton` didn't emit `container: true` on its `Semantics`. I fixed it by updating all the tests to set an appropriate theme.

Again, I'm not asking about fixing these particular bugs in the widgets, but documenting the intention of **stand-alone** vs **building-block** on every Flutter widget provided, because it's critical to building with these widgets with **correct accessibility semantics**. Too often, I've run into the UI that looks right, but has serious accessibility issues because this distinction is not clear to folks building using these widgets.

**Comment by justinmc** [comment]
@chunhtai Are you the right person for this?

**Comment by chunhtai** [comment]
yes I agree we should have more info on how different widget would participate in semantics tree, but I don't have a particular strategy to achieve this at the moment. Updating one or two widgets won't help much, we would need a more systematic solution
